From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Diego Domingos <diegodo@linux.vnet.ibm.com>
Date: Thu, 25 Aug 2022 11:37:56 -0400
Subject: [PATCH] ieee1275: implement vec5 for cas negotiation

As a legacy support, if the vector 5 is not implemented, Power
Hypervisor will consider the max CPUs as 64 instead 256 currently
supported during client-architecture-support negotiation.

This patch implements the vector 5 and set the MAX CPUs to 256 while
setting the others values to 0 (default).

Signed-off-by: Diego Domingos <diegodo@linux.vnet.ibm.com>
Signed-off-by: Robbie Harwood <rharwood@redhat.com>
(cherry picked from commit f735c65b6da8a9d4251242b37774e1a517511253)
(cherry picked from commit 1639f43b2db4ac405ac2a92e50ed4cff351c3baa)
---
 grub-core/kern/ieee1275/init.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/grub-core/kern/ieee1275/init.c b/grub-core/kern/ieee1275/init.c
index 1414695cc6..37f3098c39 100644
--- a/grub-core/kern/ieee1275/init.c
+++ b/grub-core/kern/ieee1275/init.c
@@ -307,6 +307,18 @@ struct option_vector2 {
   grub_uint8_t max_pft_size;
 } __attribute__((packed));
 
+struct option_vector5 {
+        grub_uint8_t byte1;
+        grub_uint8_t byte2;
+        grub_uint8_t byte3;
+        grub_uint8_t cmo;
+        grub_uint8_t associativity;
+        grub_uint8_t bin_opts;
+        grub_uint8_t micro_checkpoint;
+        grub_uint8_t reserved0;
+        grub_uint32_t max_cpus;
+} __attribute__((packed));
+
 struct pvr_entry {
   grub_uint32_t mask;
   grub_uint32_t entry;
@@ -325,6 +337,8 @@ struct cas_vector {
   grub_uint16_t vec3;
   grub_uint8_t vec4_size;
   grub_uint16_t vec4;
+  grub_uint8_t vec5_size;
+  struct option_vector5 vec5;
 } __attribute__((packed));
 
 /* Call ibm,client-architecture-support to try to get more RMA.
@@ -345,7 +359,7 @@ grub_ieee1275_ibm_cas (void)
   } args;
   struct cas_vector vector = {
     .pvr_list = { { 0x00000000, 0xffffffff } }, /* any processor */
-    .num_vecs = 4 - 1,
+    .num_vecs = 5 - 1,
     .vec1_size = 0,
     .vec1 = 0x80, /* ignore */
     .vec2_size = 1 + sizeof(struct option_vector2) - 2,
@@ -356,6 +370,10 @@ grub_ieee1275_ibm_cas (void)
     .vec3 = 0x00e0, // ask for FP + VMX + DFP but don't halt if unsatisfied
     .vec4_size = 2 - 1,
     .vec4 = 0x0001, // set required minimum capacity % to the lowest value
+    .vec5_size = 1 + sizeof(struct option_vector5) - 2,
+    .vec5 = {
+	0, 0, 0, 0, 0, 0, 0, 0, 256
+    }
   };
 
   INIT_IEEE1275_COMMON (&args.common, "call-method", 3, 2);
